test/cql-pytest/fetch_scylla.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/python
+# Fetch from ScyllaDB's S3 bucket (downloads.scylladb.com) a pre-compiled
+# version of Scylla for a desired release, ready to be run by tests (run.py).
+# What we need to fetch is a "relocatable" package containing just the Scylla
+# executable and the shared libraries it needs - we don't need a full OS and
+# not even stuff like JMX or Python that cql-pytest tests don't need.
+#
+# Can fetch released versions with names like:
+#   * 5.4.7
+#   * 5.4 (latest in the 5.4 branch, e.g., 5.4.7)
+#   * 5.4.0~rc2 (prerelease)
+#   * 2021.1.9 (Enterprise release)
+#   * 2023.1 (latest in this branch, Enterprise release)
+
+import boto3
+import sys
+import os
+import subprocess
+
+# Parse version 5.6.2 into the array [5, 6, 2].
+def parse_version(version):
+    return version.split('.')
+
+def s3_download(s3, bucket, object_key, out):
+    print(f'Downloading {object_key}... ', end='')
+    if not sys.stdout.isatty():
+        bucket.download_file(object_key, out)
+    else:
+        # When interactive, show download progress:
+        print('')
+        objlen = s3.Object(bucket.name, object_key).content_length
+        downloaded = 0
+        def progress(chunk):
+            nonlocal downloaded
+            downloaded += chunk
+            print('\33[2K\r%02d%%  %d MB' % (int(downloaded*100/objlen), int(downloaded/1024/1024)), end='')
+        bucket.download_file(object_key, out, Callback=progress)
+        print('')
+    print('done.')
+
+# Download Scylla release "release" (e.g., 5.4, 5.4.7, 2023.1, 2023.1.7)
+# into a subdirectory of directory "dir" named by the specific release
+# downloaded (e.g., 5.4.7) and return the ready-to-use scylla executable
+# (actually, a shell-script wrapper) in that directory - or raise exception
+# if this release cannot be found.
+# If the same download was already done in the past, discover this as quickly
+# as possible (a 3-component release like 5.4.7 doesn't even require a network
+# lookup, a 2-component release like 5.4 does) and don't download again.
+def download_scylla(release, dir):
+    scylla_arch = 'x86_64'
+
+    v = parse_version(release)
+    # If release has 3 components (e.g., 5.4.7) we know exactly which version
+    # we are looking for, and can check if we already downloaded it without
+    # even fetching the list of versions from s3:
+    if len(v) == 3 and os.path.exists(dir + '/' + release + '/scylla_wrapper'):
```

Comment:
does it compare the real version string or provided arg? Best would be to cache on exact version (like when asking for 5.4, gets 5.4.1 but later should get 5.4.2 correctly.

Reply:
Yes (I explained this somewhere but perhaps not in this specific comment):
If you ask for a release number with 3 components, like 5.4.7, the code can check immediately if we already have this version cached and finish immediately.
But if the release number has 2 components, like 5.4, the code needs to continue, it needs to look in s3 which releases we have in 5.4, find the latest one (e.g., 5.4.7), and then the code checks **again** maybe we have 5.4.7 (look for the "Already downloaded" string).